src/matrix.c

```diff
@@ -0,0 +1,250 @@
+    R_len_t endcol = startcol + chunksize - 1;
+    if (endcol >= dtncol) 
+      endcol = dtncol - 1;
+    
+    // Determine where to fill in the matrix vector
+    R_xlen_t vecIdx = startcol * n;
+    
+    // Iterate through columns in DT, copying the contents to the matrix column
+    for (R_len_t jj = startcol; jj <= endcol; jj++) {
+      memcpy(pmat + vecIdx, pcol[jj], sizeof(Rcomplex)*n);
+      vecIdx += n;
+    }
+  }
+
+UNPROTECT(1);
+return mat;
+}
+
+SEXP asmatrix_character(SEXP dt, R_xlen_t matlen, R_len_t n) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(STRSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_STRING_ELT(mat, vecIdx, STRING_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+SEXP asmatrix_list(SEXP dt, R_xlen_t matlen, R_len_t n) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(VECSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_VECTOR_ELT(mat, vecIdx, VECTOR_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+// Dispatch function for different atomic types
+SEXP asmatrix(SEXP dt) {
+  // Determine the number of rows and columns in the data.table
+  R_xlen_t p = xlength(dt);
+  R_xlen_t n = xlength(VECTOR_ELT(dt, 0));
+  
+  // Determine the number of elements in the resulting matrix
+  R_xlen_t matlen = n * p;
+  
+  /* Conversion to a common atomic type is handled in R. We detect the
+   * atomic type from the first column. */
+  SEXPTYPE R_atomic_type = TYPEOF(VECTOR_ELT(dt, 0));
+  
+  // Copy values from dt into the matrix mat using appropriately typed function
+  switch(R_atomic_type) {
+    case LGLSXP: 
+      return(asmatrix_logical(dt, matlen, n));
+    case INTSXP: 
+      return(asmatrix_integer(dt, matlen, n));
+    case REALSXP: 
+      return(asmatrix_numeric(dt, matlen, n));
+    case CPLXSXP: 
+      return(asmatrix_complex(dt, matlen, n));
+    case STRSXP: 
+      return(asmatrix_character(dt, matlen, (R_len_t) n));
+    case VECSXP:
+      return(asmatrix_list(dt, matlen, (R_len_t) n));
+    default:
+      error("Unsupported matrix type '%s'", type2char(R_atomic_type));
```

Comment:
this is untested, is it possible to get here? `expression` and `raw` inputs are my go to for triggering such case

Reply:
If you think it's impossible, please mark as `Internal error` and `// # nocov`